### PR TITLE
NEP-35 support for slice_with_int_dask_array

### PR DESCRIPTION
**Pull request by pentschev**
Adds NEP-35 support for `slice_with_int_dask_array`, as well as CuPy tests to verify it works as expected.

### CODE REVIEW

**Comment by pentschev** [comment]
This is not a high priority PR and it's not necessary to make today's Dask release.

**Comment by pentschev** [comment]
Does anyone know if something is going on with codecov? I don't think this PR is really decreasing coverage by 73%. 🤣

File: dask/array/tests/test_cupy.py [hunk 1]
```diff
@@ -1468,3 +1468,120 @@ def test_cupy_cholesky(shape, chunk):
         check_graph=False,
         check_chunks=False,
     )
+
+
```
Comment (quasiben) [inline dask/array/tests/test_cupy.py:1472]: Can you also add the `@pytest.mark.gpu` line.  xref https://github.com/dask/dask/pull/7876 for more details

**Comment by quasiben** [comment]
I see codcov reporting 93.50% coverage

**Comment by pentschev** [comment]
> I see codcov reporting 93.50% coverage

Weird, when I wrote I saw it reporting -73%, probably a glitch.

File: dask/array/tests/test_cupy.py (hunk 1 above)
Comment (pentschev) [reply dask/array/tests/test_cupy.py:1472]: I don't know for sure, but it seems like https://github.com/dask/dask/blob/main/dask/array/tests/test_cupy.py#L5 is marking the whole file as GPU. No single tests have that mark in that file. Could you confirm @charlesbluca ?

Comment (charlesbluca) [reply dask/array/tests/test_cupy.py:1472]: Yup, this entire file is marked for GPU testing

Comment (quasiben) [reply dask/array/tests/test_cupy.py:1472]: Ah!  Sorry for the noise then

**Comment by quasiben** [comment]
Dask is releasing today: https://github.com/dask/community/issues/171.  Can this wait until after the release ?

**Comment by pentschev** [comment]
> Dask is releasing today: [dask/community#171](https://github.com/dask/community/issues/171). Can this wait until after the release ?

That's what I wrote in https://github.com/dask/dask/pull/7927#issuecomment-885632625 , it can wait.

**Comment by jsignell** [comment]
>> Weird, when I wrote I saw it reporting -73%, probably a glitch.

Codecov comes in as the tests finish. So it always shows really low until the tests with all the deps run.

**Comment by pentschev** [comment]
> Codecov comes in as the tests finish. So it always shows really low until the tests with all the deps run.

Oh, I didn't know that! Thanks Julia for the information! 😄

**Comment by quasiben** [comment]
Thanks again @pentschev .  With `2021.07.1 ` released, I'll merge now

**Comment by pentschev** [comment]
Thanks @quasiben !

**Pull request merged**